### DISCUSSION TITLE
Interop with Direct3D 11

### DISCUSSION THREAD

**Discussion by Tape-Worm** [AvaloniaUI/Avalonia#5432] [Q&A] [answered]
I have written a [graphics API](https://github.com/Tape-Worm/Gorgon) that uses Direct3D 11.x (via SharpDX), and currently it uses Windows Forms to interact with the swap chain. This API also has an editor application that is entirely built in Windows Forms. Because of this, I'm tied to .NET Framework 4.x.  I am aware that .NET 5/Core 3.1 "supports" WPF/WinForms, but that support is... well... it's crap so far (at least where designers are concerned).

So, I would like to jump to .NET 5 and I've done some preliminary work on this and the performance gains were impressive even with no code changes. However the UI stuff has proven to be a painful challenge. I _could_ jump to WPF and I have a WPF interop layer, but there are limitations that are imposed by WPF, and it introduces stuttering when rendering. It's not ideal at all.  So, I'm looking into other UI frameworks and Avalonia comes up a lot as a good alternative for UI on .NET 5/Core.

So, to my question: Am I able to integrate Direct3D with Avalonia? More specifically, can I assign a swap chain to a control? 

Just as a note: I have dug through the repo and I saw an interop sample there, but from the issues list, I saw that it no longer works properly ([Issue #2024](https://github.com/AvaloniaUI/Avalonia/issues/2024)). But that was so long ago that things may have changed?

TLDR:
Can I interop Direct3D 11.x with Avalonia?

**Comment by kekekeks** [comment, marked as answer]
>Can I interop Direct3D 11.x with Avalonia?

Short answer: yes, that's possible, but requires tinkering with Skia and ANGLE
Long answer: 
1) configure `Win32PlatformOptions` to always enable GPU and `AngleOptions` to use DirectX11
2) implement `ICustomDrawingOperation`, that will give you access to Skia on the *render thread*, see [CustomSkiaPage.cs](https://github.com/AvaloniaUI/Avalonia/blob/master/samples/RenderDemo/Pages/CustomSkiaPage.cs) example. Note that the following steps can be ONLY done from ICustomDrawingOperation::Draw callback.
3) import your Direct3D content from DXGI share handle, wrap it into EGLSurface, bind EGLSurface to a texture, import said texture to Skia
4) draw imported image on SKCanvas
5) cleanup

Obviously the 3rd step is the hardest one.
```cs
var egl = (EglPlatformOpenGlInterface)AvaloniaLocator.Current.GetService<IPlatformOpenGlInterface>();
var angle = (AngleWin32EglDisplay)egl.Display;
```

This will give you access to EGL and OpenGL entry points. `AngleWin32EglDisplay` already has `GetDirect3DDevice` and `WrapDirect3D11Texture` methods. EGLSurface->texture binding can be done via [eglBindTexImage](https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglBindTexImage.xhtml). 


At some point I'll make it work out of the box with some simple D3DImage-like API, but that probably won't happen until summer/fall.

**Reply by kekekeks** [reply]
Alternatively you can just use NativeControlHost and embed a proper HWND.

**Reply by Tape-Worm** [reply]
Awesome, thank you.  

Are there any airspace issues with NativeControlHost?

**Reply by kekekeks** [reply]
>Are there any airspace issues with NativeControlHost?

Yes, you can't have any Avalonia content on top of the embedded HWND

**Reply by Tape-Worm** [reply]
That's what I figured.

Using the Skia/ANGLE approach, are there any limitations?  For example, in WPF, using D3DImage, you're limited to 60 frames per second (and, when I maximize the window 30 FPS...), 

I figure I'll probably use the same approach that I used with WPF and create a wrapper to interop with it using Skia/ANGLE.

**Reply by kekekeks** [reply]
>are there any limitations

ICustomDrawingOperation::Render is called using the same rate as the rest of the renderer (as long as you invalidate your visual), so FPS should be fine. Note that proper render timer is only available on Windows 10 with WinUI composition enabled.

**Reply by Tape-Worm** [reply]
OK, that sounds fine.

Just a note, when I set UseWinUIComposition to true, I get a warning in my output:
"[WinUIComposition] Unable to initialize WinUI compositor: Windows 10 Build 17134 is required. Your machine has Windows 10 Build 19042 installed."

Just thought you should know.

**Reply by Tape-Worm** [reply]
```csharp
var egl = (EglPlatformOpenGlInterface)AvaloniaLocator.Current.GetService<IPlatformOpenGlInterface>();
var angle = (AngleWin32EglDisplay)egl.Display;
```

Is not working for me.  The `egl` value is always null.  I'm not sure if I've got this configured correctly.  

Here's my setup code:
```csharp
          => AppBuilder.Configure<App>()
                .UsePlatformDetect()
                .With(() => new Win32PlatformOptions
                {
                    UseWindowsUIComposition = true
                })
                .With(() => new AngleOptions
                {
                    AllowedPlatformApis =
                    {
                        AngleOptions.PlatformApi.DirectX11
                    }                    
                })
                .UseSkia()
                .LogToTrace();
```
Is there anything there that might be incorrect?

Also, I'm not entirely sure how I'm supposed to convert an EglSurface to a SkSurface so I can render it?

**Reply by kekekeks** [reply]
Try adding `AllowEglInitialization = true`. It should be enabled automatically though.

**Reply by Tape-Worm** [reply]
OK, I think this might be the reason:
```
[OpenGL] Unable to initialize EGL-based rendering: 'System.DllNotFoundException: Unable to load DLL 'av_libGLESv2.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
   at Avalonia.OpenGL.Angle.AngleEglInterface.EGL_GetProcAddress(String proc)
   at Avalonia.OpenGL.Angle.AngleEglInterface.LoadAngle() in /_/src/Avalonia.OpenGL/Angle/AngleEglInterface.cs:line 21
   at Avalonia.OpenGL.Angle.AngleEglInterface..ctor() in /_/src/Avalonia.OpenGL/Angle/AngleEglInterface.cs:line 12
   at Avalonia.OpenGL.Angle.AngleWin32EglDisplay..ctor() in /_/src/Avalonia.OpenGL/Angle/AngleWin32EglDisplay.cs:line 65
   at Avalonia.Win32.Win32GlManager.<>c.<Initialize>b__1_1() in /_/src/Windows/Avalonia.Win32/Win32GlManager.cs:line 27
   at Avalonia.OpenGL.Egl.EglPlatformOpenGlInterface.TryCreate(Func`1 displayFactory) in /_/src/Avalonia.OpenGL/Egl/EglPlatformOpenGlInterface.cs:line 34'
```
Pops up when I run the application (I hadn't noticed it yesterday - my bad).

**Reply by kekekeks** [reply]
Please check that `Avalonia.Angle.Windows.Natives ` package is installed and that av_libGLESv2.dll is loadable. You might need to manually copy it to your bin directory if you are targeting the legacy .NET framework.

**Reply by Tape-Worm** [reply]
Ok, that seemed to do it.  I had to copy the native library manually, and I had to remove the AngleOptions from my initialization code.  I now have an object to work with.

**Reply by Tape-Worm** [reply]
OK, I was able to create the EglSurface from my render target.  But I can't seem to find anything related to EglBindTexImage in the code for Avalonia.

 I can't claim to be too familiar with OpenGL, so it's not immediately obvious what I'm supposed to do with this surface and skia.  

Have I missed or misunderstood something?

**Reply by kekekeks** [reply]
We do not provide full OpenGL or EGL API bindings. You'll need to use `GetProcAddress` from `EglInterface` to obtain the function pointer and use `Marshal.GetDelegateFromFunctionPointer` to get a delegate.

**Reply by Tape-Worm** [reply]
Fair enough.  Thanks for all your help.

**Reply by Tape-Worm** [reply]
Before I forget, the `WrapDirect3D11Texture` method did not work for me.  I've added an issue for it (#5455).

**Reply by amerkoleci** [reply]
@Tape-Worm Did you managed to handle the D3D11 part? When do you call SwapChain.Present?

**Reply by Tape-Worm** [reply]
@amerkoleci I never got this working unfortunately. And I wasn't using a swap chain for this, I was just updating a render target texture.
